File: DsDotNet/src/Engine/Engine.Parser/3.2.ElementListener.cs
using System.Xml.Linq;

using static Engine.Core.GraphModule;
using static Engine.Core.TextUtil;

namespace Engine.Parser;


/// <summary>
/// Alias map 및 Api map 가 생성된 이후의 처리
/// </summary>
class ElementListener : ListenerBase
{
    public ElementListener(dsParser parser, ParserHelper helper)
        : base(parser, helper)
    {
    }


    override public void EnterParenting(ParentingContext ctx)
    {
        base.EnterParenting(ctx);
    }

    public override void EnterInterfaceDef(InterfaceDefContext ctx)
    {
        var hash = _system.ApiItems;
        var interrfaceNameCtx = findFirstChild<InterfaceNameContext>(ctx);
        var interfaceName = collectNameComponents(interrfaceNameCtx)[0];
        string[][] collectCallComponents(CallComponentsContext ctx) =>
            enumerateChildren<Identifier123Context>(ctx)
                .Select(collectNameComponents)
                .ToArray()
                ;
        bool isWildcard(string[] cc) => cc.Length == 1 && cc[0] == "_";
        Real[] findSegments(string[][] fqdns) =>
            fqdns
            .Where(fqdn => fqdn != null)
            .Select(s => _model.FindGraphVertex<Real>(s))
            .Tap(x => Assert(x != null))
            .ToArray()
            ;
        var ser =   // { start ~ end ~ reset }
            enumerateChildren<CallComponentsContext>(ctx)
            .Select(collectCallComponents)
            .Tap(callComponents => Assert(callComponents.ForAll(cc => cc.Length == 2 || isWildcard(cc))))
            .Select(callCompnents => callCompnents.Select(cc => isWildcard(cc) ? null : cc.Prepend(_system.Name).ToArray()).ToArray())
            .ToArray()
            ;
        var item = hash.First(it => it.Name == interfaceName);
        var n = ser.Length;

        Assert(n == 2 || n == 3);
        item.AddTXs(findSegments(ser[0]));
        item.AddRXs(findSegments(ser[1]));

        Console.WriteLine();
    }
    override public void EnterCausalToken(CausalTokenContext ctx)
    {
        var ns = collectNameComponents(ctx);
        Assert(ns.Length.IsOneOf(1, 2));

        var path = AppendPathElement(ns);

        var existing = _modelSpits.Where(spit => spit.NameComponents.IsStringArrayEqaul(path)).ToArray();
        if (existing.Where(spit => spit.GetCore() is Vertex).Any())
            return;

        var pathWithoutParenting = new[] { _system.Name, _flow.Name }.Concat(ns).ToArray();

        // narrow match
        var matches =
            _modelSpits
            .Where(spitResult =>
                spitResult.NameComponents.IsStringArrayEqaul(path)
                || spitResult.NameComponents.IsStringArrayEqaul(pathWithoutParenting))
            .Select(spitResult => spitResult.GetCore())
            .ToArray()
            ;

        var pathAdapted = ns.Length == 2 ? new[] { _system.Name }.Concat(ns).ToArray() : new string[] { };

        // 나의 시스템의 다른 flow 에 존재하는 segment 호출
        var extendedMatches =
            _modelSpits
            .Where(spitResult =>
                pathAdapted.Any() && spitResult.NameComponents.IsStringArrayEqaul(pathAdapted))
            .Select(spitResult => spitResult.GetCore())
            .ToArray()
            ;

        // 다른 시스템의 API 호출
        var apiCall =
            _modelSpitObjects
                .OfType<ApiItem>()
                .Where(api => api.NameComponents.IsStringArrayEqaul(ns))
                .FirstOrDefault();

        Assert(matches.Length.IsOneOf(0, 1));

        // API call 과 나의 시스템의 다른 flow 에 존재하는 segment 호출이 헷갈리지 않도록
        if (extendedMatches.OfType<Real>().Any(r => r.NameComponents.IsStringArrayEqaul(pathAdapted)))
        {
            if (apiCall != null)
                throw new ParserException($"Ambiguous entry [{apiCall.QualifiedName}] and [{pathAdapted.Combine()}]", ctx);
            Alias.CreateInFlow(ns.Combine(), pathAdapted, _flow, true);
            return;
        }

        if (matches.OfType<Real>().Any())
            return;

        try
        {
            var alias = matches.OfType<SpitOnlyAlias>().FirstOrDefault();
            if (alias != null)
            {
                var aliasKey =
                    matches
                        .OfType<SpitOnlyAlias>()
                        .Where(alias => alias.Mnemonic.IsStringArrayEqaul(pathWithoutParenting))
                        .Select(alias => alias.AliasKey)
                        .FirstOrDefault()
                        ;
                switch (aliasKey.Length)
                {
                    case 3:     // my flow real 에 대한 alias
                        Assert(aliasKey[0] == _system.Name && aliasKey[1] == _flow.Name);
                        Alias.CreateInFlow(ns.Combine(), aliasKey, _flow);
                        return;
                    case 2:
                        var apiItem =
                            _modelSpitObjects
                                .OfType<ApiItem>()
                                .Where(api => api.NameComponents.IsStringArrayEqaul(aliasKey))
                                .FirstOrDefault();
                        Assert(apiItem != null);

                        if (_parenting == null)
                            Alias.CreateInFlow(ns.Combine(), aliasKey, _flow);
                        else
                            Alias.CreateInReal(ns.Combine(), apiItem, _parenting);
                        return;
                    case 1:
                        Assert(false);
                        break;
                }
            }


            if (apiCall != null)
            {
                if (_parenting == null)
                    Call.CreateInFlow(apiCall, _flow);
                else
                    Call.CreateInReal(apiCall, _parenting);
                return;
            }


            var prop = _elements[path];
            if (_parenting == null)
            {
                if(ns.Length != 1)
                    throw new ParserException($"ERROR: unknown token [{ns.Combine()}].", ctx);
                Real.Create(ns[0], _flow);
                return;
            }
            else
                throw new ParserException($"ERROR: unknown token [{ns.Combine()}].", ctx);
        }
        finally
        {
            UpdateModelSpits();
        }

    }

    override public void EnterIdentifier12Listing(Identifier12ListingContext ctx)
    {
        // side effects
        var path = AppendPathElement(collectNameComponents(ctx));
        var prop = _elements[path];
        if (_parenting != null)
            throw new ParserException($"ERROR: identifier [{path.Combine()}] not allowed!", ctx);

        Real.Create(path.Last(), _flow);
    }
}

File: DsDotNet/src/Engine/Engine.Parser.FS/3.2.ElementListener.fs
namespace Engine.Parser.FS

open System.Linq

open Engine.Common.FS
open Engine.Parser
open Engine.Core
open type Engine.Parser.dsParser
open type Engine.Parser.FS.DsParser

#if false

/// <summary>
/// Alias map 및 Api map 가 생성된 이후의 처리
/// </summary>
type ElementListener(parser:dsParser, helper:ParserHelper) =
    inherit ListenerBase(parser, helper)

    //override x.EnterAliasListing(ctx:AliasListingContext) =
    //    let flow = x._flow.Value
    //    let map = flow.AliasDefs
    //    let mnemonics = Descendants<AliasMnemonicContext>(ctx).Select(getText).ToArray()
    //    let aliasTargetCtx = TryFindFirstChild<AliasDefContext>(ctx).Value    // {타시스템}.{interface} or {Flow}.{real}
    //    let aliasTargetName = TryGetName(aliasTargetCtx).Value
    //    let realTargetCandidate = flow.Graph.TryFindVertex<Real>(aliasTargetName)
    //    let callTargetCandidate = tryFindCall helper.TheSystem aliasTargetName
    //    let target =
    //        match realTargetCandidate, callTargetCandidate with
    //        | Some real, None -> RealTarget real
    //        | None, Some call -> CallTarget call
    //        | Some _, Some _ -> failwith "Name duplicated."
    //        | _ -> failwith "Failed to find"

    //    { AliasTarget=target; Mnemonincs=mnemonics} |> map.Add

    //    let ci = getContextInformation aliasTargetCtx
    //    x.AddElement(ci, GVT.AliaseKey)

    //    //for mne in mnemonics do
    //    //    x.AddElement(mne, GVT.AliaseMnemonic)
    //    //let aliasesHash = mnemonics.Select(fun ctx -> ctx.Names.Combine()).ToHashSet()
    //    //let aliasKey = ci.Names.ToArray()
    //    //map.Add(aliasKey, aliasesHash)



    override x.EnterInterfaceDef(ctx:InterfaceDefContext) =
        let system = helper.TheSystem
        let hash = system.ApiItems4Export
        let interrfaceNameCtx = TryFindFirstChild<InterfaceNameContext>(ctx)
        let interfaceName = CollectNameComponents(interrfaceNameCtx.Value)[0]

        let collectCallComponents(ctx:CallComponentsContext):Fqdn[] =
            Descendants<Identifier123Context>(ctx)
                .Select(CollectNameComponents)
                .ToArray()

        let isWildcard(cc:Fqdn):bool = cc.Length = 1 && cc[0] = "_"
        let findSegments(fqdns:Fqdn[]):Real[] =
            fqdns
                .Where(fun fqdn -> fqdn <> null)
                .Select(fun s -> system.FindGraphVertex<Real>(s))
                .Tap(fun x -> assert(not (isItNull x)))
                .ToArray()

        let ser =   // { start ~ end ~ reset }
            Descendants<CallComponentsContext>(ctx)
                .Select(collectCallComponents)
                .Tap(fun callComponents -> assert(callComponents.All(fun cc -> cc.Length = 2 || isWildcard(cc))))
                .Select(fun callCompnents -> callCompnents.Select(fun cc -> if isWildcard(cc) then null else cc.Prepend(system.Name).ToArray()).ToArray())
                .ToArray()

        let item = hash.First(fun it -> it.Name = interfaceName)
        let n = ser.Length

        assert(n = 2 || n = 3)
        item.AddTXs(findSegments(ser[0])) |> ignore
        item.AddRXs(findSegments(ser[1])) |> ignore


    override x.EnterCausalToken(ctx:CausalTokenContext) =
        let ci = getContextInformation ctx
        let sysNames, flowName, parenting, ns = ci.Tuples
        let flow = x._flow.Value
        let system = helper.TheSystem
        let spits = system.Spit()
        let findSpits (ns:string seq) =
            spits.Where(fun sp -> sp.NameComponents = (ns.ToArray()) || sp.NameComponents.Combine() = ns.Combine()).ToArray()
        let findSpit ns = findSpits ns |> Array.tryHead

        let createRealTargetAlias (ci:ContextInformation) (target:Real) =
            let aliasCreator =
                let aliasTarget = new AliasTargetReal(target.NameComponents)
                let parent = choiceParentWrapper ci x._flow x._parenting
                new AliasCreator(ci.Names.Combine(), parent, aliasTarget)
            x.ParserHelper.AliasCreators.Add(aliasCreator)

        let createAlias(soa:SpitOnlyAlias) =
            let aliasKey = soa.AliasKey
            match aliasKey.Length with
                | 1 ->     // my flow real 에 대한 alias
                    let aliasCreator =
                        let aliasTarget = new AliasTargetReal([|yield! soa.FlowFqdn; yield! aliasKey|])
                        new AliasCreator(ns.Combine(), Flow flow, aliasTarget)
                    x.ParserHelper.AliasCreators.Add(aliasCreator) |> ignore
                | 2 ->
                    let apiItem =
                        system.ApiItems.Where(fun api -> api.NameComponents = aliasKey)
                            .Head()

                    let name = ns.Combine()
                    match x._parenting with
                    | Some parent ->
                        let aliasCreator =
                            let aliasTarget = new AliasTargetApi(apiItem)
                            new AliasCreator(name, Real parent, aliasTarget)
                        x.ParserHelper.AliasCreators.Add(aliasCreator)
                    | None ->
                        (* flow 바로 아래에 사용되는 직접 call.  A.+ *)
                        let aliasCreator =
                            let aliasTarget = new AliasTargetDirectCall(aliasKey)
                            new AliasCreator(name, Flow flow, aliasTarget)
                        x.ParserHelper.AliasCreators.Add(aliasCreator)
                    |> ignore
                | _ ->
                    failwith "ERROR"


        let createCall (vertexType:GVT) (ci:ContextInformation) : Indirect =
            assert(ci.System.Value = system.Name)
            assert(ci.Flow.Value = flow.Name)

            let parentWrapper = tryFindParentWrapper system ci
            let existing =
                option {
                    let! parent = parentWrapper
                    return! parent.GetGraph().TryFindVertex(ci.GetRawName())
                }
            let existingOk =
                match vertexType, existing with
                | HasFlag GVT.CallFlowReal, Some (:? VertexOtherFlowRealCall as v) -> Some (v :> Indirect)
                | HasFlag GVT.CallAliased,  Some (:? VertexAlias as v) ->  Some (v :> Indirect)
                | HasFlag GVT.Call,         Some (:? VertexCall as v) ->  Some (v :> Indirect)
                | _ -> None
            match existingOk with
            | Some v -> v
            | None when existing.IsSome -> failwith "ERROR"     // 이름 동일하고 다른 type 의 vertex 가 존재하는 경우
            | None ->
                let newVertex =
                    option {
                        let! parent = parentWrapper
                        match vertexType with
                        | HasFlag GVT.CallFlowReal ->
                            match ci.Tuples with
                            | Some s, Some f, None, ofName::ofrName::[] ->   // "ofr" Other Flow Real
                                let! ofr = tryFindReal system ofName ofrName
                                return VertexOtherFlowRealCall.Create(ofName, ofrName, ofr, Flow flow) :> Indirect
                            | _ ->
                                failwith "ERROR"

                        | HasFlag GVT.CallAliased ->
                            let name = ci.GetRawName()
                            let! target = tryFindAliasTarget flow name
                            return VertexAlias.Create(name, target, parent) :> Indirect

                        | HasFlag GVT.Call ->
                            match ci.Tuples with
                            | Some s, Some f, parenting_, callName::[] ->
                                let! call = tryFindCall system callName
                                return VertexCall.Create(callName, call, parent) :> Indirect
                            | _ ->
                                failwith "ERROR"
                        | _ ->
                            failwith "ERROR"
                    }
                newVertex.Value

                //match ci.Tuples with
                //| Some s, Some f, _, device::api::[] ->     // my / flow / (parenting) / device.api
                //    match tryFindImportApiItem system [device; api] with
                //    | Some apiItem ->
                //        let parent = choiceParentWrapper ci x._flow x._parenting
                //        Call.Create(apiItem, parent) |> ignore
                //    | None ->
                //        match (findSpit [s; device; api]).OrElse(findSpit [device; api]) with
                //        | Some sp ->
                //            match sp.SpitObj with
                //            | SpitOnlyAlias soa -> createAlias soa
                //            | SpitReal real -> createRealTargetAlias ci real
                //            | _ -> failwith "Not an API item"
                //        | None -> tracefn "Need to generate %A" [s; device; api]
                //| _ ->
                //    failwith "ERROR"

        let createAliasFromContextInformation (ci:ContextInformation) =
            let alias =
                (findSpits ci.Names)
                    .Select(fun sp -> sp.GetCore())
                    .OfType<SpitOnlyAlias>()
                    .Where(fun sp -> sp.FlowFqdn = [| ci.System.Value; ci.Flow.Value |] )
                    .ExactlyOne()

            createAlias alias



        let vertexType = helper._causalTokenElements[ci]
        match vertexType with
        | HasFlag GVT.Segment ->
            let existing = findSpit(ci.NameComponents)
            match existing, ns with
            | Some _,   _ -> ()
            | None,     _::_::[] -> createCall vertexType ci |> ignore
            | _ -> failwith "ERROR"
        | HasFlag GVT.AliaseMnemonic ->
            createAliasFromContextInformation ci
        | ( HasFlag GVT.CallAliased
          | HasFlag GVT.Call
          | HasFlag GVT.CallFlowReal
          | HasFlag GVT.Child ) ->
            createCall vertexType ci |> ignore
        | _ ->
            failwith "ERROR"
        x.UpdateModelSpits()
#endif